rs/canister_sandbox/src/protocol/structs.rs

```diff
@@ -36,20 +36,42 @@ pub struct SandboxExecInput {
 pub struct SandboxExecOutput {
     pub slice: SliceExecutionOutput,
     pub wasm: WasmExecutionOutput,
-    pub state: Option<StateModifications>,
+    pub state: StateModifications,
     pub execute_total_duration: std::time::Duration,
     pub execute_run_duration: std::time::Duration,
 }
 
+impl SandboxExecOutput {
+    pub fn take_state_modifications(&mut self) -> StateModifications {
+        std::mem::take(&mut self.state)
+    }
+}
+
 /// Describes the memory changes performed by execution.
 #[derive(Clone, PartialEq, Debug, Deserialize, Serialize)]
 pub struct MemoryModifications {
     pub page_delta: PageDeltaSerialization,
     pub size: NumWasmPages,
 }
 
-#[derive(Clone, PartialEq, Debug, Deserialize, Serialize)]
+#[derive(Serialize, Default, Debug, Deserialize, Clone, PartialEq)]
 pub struct StateModifications {
+    /// Modifications in the execution state of the canister.
+    ///
+    /// This field is optional because the state changes might or might not
+    /// be applied depending on the method executed.
+    pub execution_state_modifications: Option<ExecutionStateModifications>,
+
+    /// Modifications in the system state of the canister.
+    ///
+    /// The system state changes contain parts that are always applied
+    /// and parts that are only applied depending on the method executed
+    /// (similarly to `execution_state_modifications`).
+    pub system_state_changes: SystemStateChanges,
```

Comment:
nit: maybe, rename `SystemStateChanges` to `SystemStateModifications` for consistency? Or vice-versa, `StateModifications` to `StateChanges`...

Reply:
Yeah I was thinking the same but I think I'll do it in a follow up to avoid extra bloat here.